### PR TITLE
Angular and Vue storyshots

### PR DESCRIPTION
**Pull request by igor-dv**
Issue: Adding storyshots support for angular

## What I did
1. I have a BDay - this is my cake =) 🍰 
2. I've refactored storyshots to make it easier to integrate new frameworks
3. There are these PRs (#2517, #2413) that should happen before
4. ~~I have concerns about adding every new framework to peer dep... Should people install all these angular/polymer/aurelia packages?~~
5. There is a lot of copy-paste - **TBD in a sperate PR - extract common angular logic to some kind of `@storybook/angular-core`**
6. ~~I don't know how to make props to be visible in snapshots =(~~
7. ~~One storyshot is failing. We need to refactor more things to be able to run framework-specific preparations with beforeEach.~~

## How to test
For running only angular sthoryshots:
```
node node_modules\jest\bin\jest.js angularshots
```
 
For running only vue sthoryshots:
```
node node_modules\jest\bin\jest.js vueshots
```

### CODE REVIEW

**Comment by igor-dv** [comment]
@storybooks/team , I think it's ready for review.

CC: @ralzinov @alterx - angular part

**Comment by ndelangen** [comment]
I'd assume this would increase our test coverage more? Is it not reported correctly yet because of failing CI?

**Comment by igor-dv** [comment]
Idk, how test coverage is calculated?

**Comment by igor-dv** [comment]
Any ideas why cli is failing? I know it runs with angular v4, but I checked it locally with v4 and storyshots did work...

**Comment by Hypnosphi** [comment]
Here's the error message, for reference:

```
ERROR in /tmp/storybook/node_modules/@angular/platform-browser-dynamic/src/compiler_factory.d.ts
(9,115): error TS2305: Module '"/tmp/storybook/node_modules/@angular/compiler/compiler"' has no exported member 'CompileReflector'.
```

**Comment by igor-dv** [comment]
Yeah, I've seen it. I but I can't understand what is this dependencies hell again.

**Comment by igor-dv** [comment]
Just moved angular deps in storyshots back to peer and cli tests are green now, FML. 
@Hypnosphi, any suggestions? I feel it's the monorepo thingy, but I'm not familiar with cli tests to be sure.

**Comment by Hypnosphi** [comment]
> moved angular deps in storyshots back to peer

Why did you need them as direct deps?

**Comment by igor-dv** [comment]
They are used in the code

**Comment by Hypnosphi** [comment]
Is there a chance someone uses storyshots without storybook?

**Comment by Hypnosphi** [comment]
See https://github.com/storybooks/storybook/pull/2467#issuecomment-355014111

**Comment by igor-dv** [comment]
Hah, I don't think so. Do you mean to remove them at all? In this case, we will get the extraneous deps warning...

**Comment by Hypnosphi** [comment]
As a react user, I'd be quite surprised if I knew that some developer tool installs angular for me

**Comment by Hypnosphi** [comment]
> In this case, we will get the extraneous deps warning...

I see two solutions here:
— Split storyshots into three different framework-specific packages, each of which has their own peerDeps
— Suppress the warnings

**Comment by igor-dv** [comment]
Yeah. I'll suppress it... Agree with you.

File: addons/storyshots/package.json [hunk 1]
```diff
@@ -20,8 +20,11 @@
     "babel-runtime": "^6.26.0",
     "glob": "^7.1.2",
     "global": "^4.3.2",
+    "jest-preset-angular": "^5.0.0",
     "jest-specific-snapshot": "^0.3.0",
     "prop-types": "^15.6.0",
+    "react": "*",
+    "react-test-renderer": "*",
```
Comment (Hypnosphi) [inline addons/storyshots/package.json]: Please don't use wildcards for direct dependencies

File: addons/storyshots/package.json [hunk 2]
```diff
@@ -20,8 +20,10 @@
     "babel-runtime": "^6.26.0",
     "glob": "^7.1.2",
     "global": "^4.3.2",
+    "jest-preset-angular": "^5.0.0",
     "jest-specific-snapshot": "^0.3.0",
     "prop-types": "^15.6.0",
+    "react-test-renderer": "^16.2.0",
```
Comment (Hypnosphi) [inline addons/storyshots/package.json]: Can you please check that it works in a react 15 app?

Comment (ndelangen) [reply addons/storyshots/package.json]: Would this block the merge? I presume this is the same for master?

Comment (Hypnosphi) [reply addons/storyshots/package.json]: On master, it's a peer dep

File: addons/storyshots/src/angular/renderTree.js [hunk 3]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable import/no-extraneous-dependencies */
```
Comment (Hypnosphi) [inline addons/storyshots/src/angular/renderTree.js]: Please disable it for a particular line only (`// eslint-disable-next-line`)

File: addons/storyshots/package.json (hunk 2 above)
Comment (igor-dv) [reply addons/storyshots/package.json]: Maybe I can just limit it to the >= 15?

File: addons/storyshots/src/angular/renderTree.js [hunk 4]
```diff
@@ -1,8 +1,10 @@
-/* eslint-disable import/no-extraneous-dependencies */
 import AngularSnapshotSerializer from 'jest-preset-angular/AngularSnapshotSerializer';
 import HTMLCommentSerializer from 'jest-preset-angular/HTMLCommentSerializer';
+// eslint-disable-next-line
```
Comment (Hypnosphi) [inline addons/storyshots/src/angular/renderTree.js]: please disable only the specific rule (`// eslint-disable-next-line import/no-extraneous-dependencies`)

**Comment by igor-dv** [comment]
So I'm waiting for #2413 to be finished and I'll be able to solve conflicts, but will be glad to get a review anyway =)

CC. @alterx , @ralzinov  for the angular part.

File: addons/storyshots/package.json (hunk 2 above)
Comment (Hypnosphi) [reply addons/storyshots/package.json]: What do you mean? It's not a peer dep. `>=15` will just install latest 16

Comment (igor-dv) [reply addons/storyshots/package.json]: Yeah, right. I'll check it with 15 so. 🤦‍♂️

**Comment by igor-dv** [comment]
@Hypnosphi, I think there will be indeed a problem with v15 (didn't check it but have seen other issues about it). So probably I have to return the `react-test-renderer` back to peer. It doesn't make sense for angular and vue users, but removing it at all will break other imports in storyshots (that I can't remove for backward compatibility)...
Ideas?

**Comment by Hypnosphi** [comment]
> Ideas?

Separate packages for different frameworks?

**Comment by igor-dv** [comment]
🤦‍♂️  I am lazy to do this

**Comment by igor-dv** [comment]
@Hypnosphi , what do you think of using this: https://github.com/Wizcorp/codependency ?

**Comment by Hypnosphi** [comment]
Description looks a bit outdated (npm stopped installing peers quite a long time ago)

**Comment by Hypnosphi** [comment]
Probably that's because it was last updated in 2014

**Comment by igor-dv** [comment]
I think I've found a solution. Now `react-test-renderer` is loaded only when used with react storyshots. I can remove it from deps (also jest-preset-angular). Need only update the docs.

File: addons/storyshots/src/angular/renderTree.js [hunk 5]
```diff
@@ -1,4 +1,6 @@
+// eslint-disable-next-line import/no-extraneous-dependencies
```
Comment (Hypnosphi) [inline addons/storyshots/src/angular/renderTree.js]: How are people expected to get `jest-preset-angular` dependency? As contrary to `@angular/core`, it's by no means guaranteed to present in an angular project

File: addons/storyshots/src/react/renderShallowTree.js [hunk 6]
```diff
@@ -1,3 +1,4 @@
+// eslint-disable-next-line import/no-extraneous-dependencies
```
Comment (Hypnosphi) [inline addons/storyshots/src/react/renderShallowTree.js]: Same here

Comment (igor-dv) [reply addons/storyshots/src/react/renderShallowTree.js]: People who are already using Storyshots were already needed to install `react-test-renderer`, so for them, there won't be a problem. I need to add a section in Readme about these "optional peers", that people have to install by hand.
BTW, For angular and vue these deps are needed to configure jestconfig (I've already added it to readme)

Comment (Hypnosphi) [reply addons/storyshots/src/react/renderShallowTree.js]: Makes sense to me

File: addons/storyshots/README.md [hunk 7]
```diff
@@ -36,6 +38,43 @@ Usually, you might already have completed this step. If not, here are some resou
 
 > Note: If you use React 16, you'll need to follow [these additional instructions](https://github.com/facebook/react/issues/9102#issuecomment-283873039).
 
+### Configure Jest for React
+StoryShots addon for React is dependent on [react-test-renderer](https://github.com/facebook/react/tree/master/packages/react-test-renderer), but 
+doesn't install it, so you need yo install it separately.
```
[the remark this answers is not in the document]
Comment (Hypnosphi) [reply addons/storyshots/README.md]: Please add the shell command to copypaste:
```sh
npm install --save-dev react-test-renderer
```

File: addons/storyshots/README.md [hunk 8]
```diff
@@ -40,11 +40,19 @@ Usually, you might already have completed this step. If not, here are some resou
 
 ### Configure Jest for React
 StoryShots addon for React is dependent on [react-test-renderer](https://github.com/facebook/react/tree/master/packages/react-test-renderer), but 
-doesn't install it, so you need yo install it separately.
+doesn't install it, so you need to install it separately (read [here](#deps-issue) why).
```
Comment (Hypnosphi) [inline addons/storyshots/README.md]: It's a super-nit, but I prefer when text keeps making sense if you replace links with their text, that's why I don't really like links with "here" text.

Can we just make "doesn't" a link?

File: addons/storyshots/README.md [hunk 9]
```diff
@@ -75,6 +87,20 @@ module.exports = {
   moduleFileExtensions: ['vue', 'js', 'jsx', 'json', 'node'],
 };
 ```
+
+### <a name="deps-issue"></a>Why don't we install dependencies of each framework ?
+Storyshots addon is currently supporting React, Angular and Vue. Each framework needs its own packages to be integrated with Jest. We don't want people that use only React will need to bring other dependencies that do not make sense for them. 
+
+`dependancies` - will installed an exact version of the particular dep - Storyshots can work with different versions of the same framework (let's say React v16 and React v15), that have to be compatible with a version of its plugin (react-test-renderer).
```
Comment (Hypnosphi) [inline addons/storyshots/README.md]: `dependancies` — typo

**Review by Hypnosphi — APPROVED** [review]

**Comment by igor-dv** [comment]
I'm merging this one since I'm tired solving conflicts. Still didn't get a feedback on the angular part, but I think it will be easier to test in beta.

**Pull request merged**